daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/Grammar.scala

```diff
@@ -89,10 +90,19 @@ class SeqComp private (context: SchemaComponent, children: Seq[Gram])
       _.isInstanceOf[NadaParser]
     }
 
+  lazy val assertExpressionChildren = parserChildren.filter {
+    _.isInstanceOf[AssertExpressionEvaluationParser]
+  }
+
   final override lazy val parser = {
     if (parserChildren.isEmpty) new NadaParser(context.runtimeData)
     else if (parserChildren.length == 1) parserChildren.head
-    else new SeqCompParser(context.runtimeData, parserChildren.toVector)
+    else
+      new SeqCompParser(
+        context.runtimeData,
+        parserChildren.toVector,
+        assertExpressionChildren.toVector,
```

Comment:
I believe the `paserChildren` vector contains the `assertExpressionChildren` parsers, which I think means the assert parsers will be evaluated twice? Once when all the `parserChildren` are evaluated and again after the parserChildren finis and then we evaluate the asserts Parsers? Seems like parserChildren should not contain the assert children?

Reply:
I believe I have fixed this.  I didn't realize (or forgot since starting this pull request a year ago) that we internally handle discriminator expressions as an assert expression just with a discriminator flag set.

I've changed it so that now we are filtering out just the discriminator statements from the rest of the sequence child parsers and run those after the rest of the sequence body.